File: bcw4/java/Unit/Unit.java
public class Unit {
    private int damage;
    private int hitPoints;
    private int hitPointsLimit;
    private String name;

    public Unit(String name, int hp, int dmg) {
        this.name = name;
        this.hitPointsLimit = hp;
        this.hitPoints = hp;
        this.damage = dmg;
    }

    public Unit() {
        this.name = "Unit";
        this.hitPointsLimit = 10;
        this.hitPoints = 10;
        this.damage = 1;
    }

    public int getDamage() {
        return damage;
    }

    public int getHitPoints() {
        return hitPoints;
    }

    public int getHitPointsLimit() {
        return hitPointsLimit;
    }
    
    public String getName() {
        return name;
    }

    public void ensureIsAlive() throws UnitIsDeadException {
        if ( hitPoints == 0 ) {
            throw new UnitIsDeadException();
        }
    }

    public void addHitPoints(int hp) throws UnitIsDeadException {
        ensureIsAlive();

        int healedHp = hitPoints + hp;

        if ( healedHp >  hitPointsLimit ) {
            healedHp = hitPointsLimit;
        }
        hitPoints = healedHp;
    }
    
    public void takeDamage(int dmg) throws UnitIsDeadException {
        ensureIsAlive();

        hitPoints -= dmg;

        if ( hitPoints < 0 ) {
            hitPoints = 0;
            throw new UnitIsDeadException();
        }
    }

    public void attack(Unit enemy) throws UnitIsDeadException {
        enemy.takeDamage(damage);
        enemy.counterAttack(this);
    }

    public void counterAttack(Unit enemy) throws UnitIsDeadException {
        enemy.takeDamage(damage/2);
    }

    public String toString() {
        return "=========================\n" + "Unit name: " + getName() + "\nUnit hitpoints: " + getHitPoints() + "/" + getHitPointsLimit() + "\nUnit damage: " + getDamage() + "\n=========================";
    }
}

File: bcw4/Unit/Unit.cpp
#include <iostream>

#include "Unit.h"

Unit::Unit(const std::string& name, int hp, int dmg) {
    this->damage = dmg;
    this->hitPoints = hp;
    this->hitPointsLimit = hp;
    this->name = name;
}

Unit::~Unit() {

}

int Unit::getDamage() const {
	return this->damage;
}

int Unit::getHitPoints() const {
    return this->hitPoints;
}

int Unit::getHitPointsLimit() const {
    return this->hitPointsLimit;
}

const std::string& Unit::getName() const {
    return this->name;
}

void Unit::ensureIsAlive() {
    if ( hitPoints == 0 ) {
        throw UnitIsDead();
    }
}

void Unit::addHitPoints(int hp) {
    ensureIsAlive();

    int healedHp = hitPoints + hp;

    if ( healedHp >  hitPointsLimit ) {
        healedHp = hitPointsLimit;
    }
    hitPoints = healedHp;
}

void Unit::takeDamage(int dmg) {
    ensureIsAlive();

    hitPoints -= dmg;

    if ( hitPoints < 0 ) {
        hitPoints = 0;
        throw UnitIsDead();
    }
}

void Unit::attack(Unit& enemy) {
    enemy.takeDamage(damage);
    enemy.counterAttack(*this);
}

void Unit::counterAttack(Unit& enemy) {
    enemy.takeDamage(damage/2);
}

std::ostream& operator<<(std::ostream& out, const Unit& unit) {
    out << "Unit type: " << unit.getName() << std::endl;
    out << "Unit hitpoints: " << unit.getHitPoints() << "/" << unit.getHitPointsLimit() << std::endl;
    out << "Unit damage: " << unit.getDamage() << std::endl;

    return out;
}
